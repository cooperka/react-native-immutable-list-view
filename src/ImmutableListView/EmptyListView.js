import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text, ListView } from 'react-native';

import ImmutableListView from './ImmutableListView';

import styles from '../styles';
import utils from '../utils';

/**
 * A ListView that displays a single item showing that there is nothing to display.
 * Useful e.g. for preserving the ability to pull-refresh an empty list.
 */
class EmptyListView extends PureComponent {

  static propTypes = {
    // Pass through any props that ListView would normally take.
    ...ListView.propTypes,

    // ImmutableListView handles creating the dataSource, so don't allow it to be passed in.
    dataSource: PropTypes.oneOf([undefined]),

    // Make this prop optional instead of required.
    renderRow: PropTypes.func,

    emptyText: PropTypes.string,
  };

  static defaultProps = {
    ...ListView.defaultProps,

    emptyText: 'No data.',
  };

  state = {
    listData: utils.UNITARY_LIST,
  };

  componentWillMount() {
    this.setListDataFromProps(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setListDataFromProps(nextProps);
  }

  setListDataFromProps(props) {
    const { listData } = this.state;
    const { renderEmpty, renderEmptyInList, emptyText } = props;

    // Update the data to make sure the list re-renders if any of the relevant props have changed.
    this.setState({
      listData: listData.set(0, Immutable.fromJS([renderEmpty, renderEmptyInList, emptyText])),
    });
  }

  /**
   * Returns a simple text element showing the `emptyText` string.
   * This method can be overridden by passing in your own `renderRow` prop instead.
   */
  renderRow() {
    const { emptyText } = this.props;

    return (
      <Text style={styles.emptyText}>
        {emptyText}
      </Text>
    );
  }

  render() {
    const { listData } = this.state;
    const {
      renderEmpty, renderEmptyInList, renderSectionHeader, emptyText, ...passThroughProps
    } = this.props;

    return (
      <ImmutableListView
        renderRow={() => this.renderRow()}
        {...passThroughProps}
        immutableData={listData}
      />
    );
  }

}

export { EmptyListView };
