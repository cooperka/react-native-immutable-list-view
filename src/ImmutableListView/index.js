export { default } from './ImmutableListView';
export * from './EmptyListView';
