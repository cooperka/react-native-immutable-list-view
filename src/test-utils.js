/* eslint-disable import/no-extraneous-dependencies */
import Immutable from 'immutable';
import React from 'react';
import { Text, ListView } from 'react-native';
import renderer from 'react-test-renderer';
/* eslint-enable */

import ImmutableListView from './ImmutableListView';
import ImmutableVirtualizedList from './ImmutableVirtualizedList';

/**
 * Some common types of data you may want to render with ImmutableListView.
 * @see https://facebook.github.io/react-native/docs/listviewdatasource.html#constructor
 */
const data = {
  EMPTY_DATA: Immutable.List(),

  LIST_DATA: Immutable.List(['lists', 'are', 'great']),

  LIST_DATA_NESTED: Immutable.List([['so', 'are'], ['nested', 'lists']]),

  MAP_DATA_LIST_ROWS: Immutable.fromJS({
    first: ['m', 'a', 'p'],
    second: ['foo'],
    third: [],
    fourth: ['bar'],
  }),

  MAP_DATA_MAP_ROWS: Immutable.fromJS({
    first: {
      row1: 'data 1',
      row2: 'data 2',
    },
    second: {},
  }),

  SET_DATA: Immutable.Set(['one', 'two', 'three']),

  RANGE_DATA: Immutable.Range(3, 10, 3),
};

const renderers = {
  /**
   * @param {*} rowData
   */
  renderRow(rowData) {
    return <Text>{JSON.stringify(rowData)}</Text>;
  },

  // eslint-disable-next-line react/prop-types
  renderItem({ item }) {
    return <Text>{JSON.stringify(item)}</Text>;
  },

  /**
   * @param {Immutable.Iterable} sectionData
   * @param {String} category
   */
  renderSectionHeader(sectionData, category) {
    return <Text header>{`${category} (${sectionData.size} items)`}</Text>;
  },
};

const mocks = {
  /**
   * Mock ScrollView so that it doesn't contain any props when rendered by ListView.
   * This is useful for comparison between ListView and ImmutableListView.
   *
   * @returns {ImmutableListView}
   */
  getImmutableListViewWithoutProps() {
    jest.resetModules();

    // eslint-disable-next-line react/prop-types
    const mockScrollView = props =>
      React.createElement('ScrollView', {}, props.children);
    jest.doMock('ScrollView', () => mockScrollView);

    // eslint-disable-next-line global-require
    return require('./ImmutableListView').default;
  },
};

const expectors = {
  expectToMatchSnapshotWithData(immutableData, shouldRenderSectionHeaders) {
    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const tree = renderer
      .create(
        <ImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectVirtualizedToMatchSnapshotWithData(immutableData) {
    const tree = renderer
      .create(
        <ImmutableVirtualizedList
          immutableData={immutableData}
          renderItem={renderers.renderItem}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectToMatchListViewWithData(immutableData, shouldRenderSectionHeaders) {
    const MockedImmutableListView = mocks.getImmutableListViewWithoutProps();

    const dataSource = new ListView.DataSource({
      rowHasChanged: (r1, r2) => r1 !== r2,
      sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
    });

    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const immutableTree = renderer
      .create(
        <MockedImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          renderEmptyInList={null}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    const updatedDataSource = dataSource.cloneWithRows(immutableData.toJS());
    const regularTree = renderer
      .create(
        <ListView
          dataSource={updatedDataSource}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    expect(immutableTree).toEqual(regularTree);
  },
};

export { data, renderers, mocks, expectors };
