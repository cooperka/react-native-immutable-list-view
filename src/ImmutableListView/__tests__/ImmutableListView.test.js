import React from 'react';
import { InteractionManager } from 'react-native';
import renderer from 'react-test-renderer';

import ImmutableListView from '../ImmutableListView';

import { data, renderers, expectors } from '../../test-utils';

describe('ImmutableListView', () => {
  it('renders with empty data', () => {
    expectors.expectToMatchSnapshotWithData(data.EMPTY_DATA);
  });

  it('renders basic List', () => {
    expectors.expectToMatchSnapshotWithData(data.LIST_DATA);
  });

  it('renders nested List', () => {
    expectors.expectToMatchSnapshotWithData(data.LIST_DATA_NESTED);
  });

  it('renders Map: List rows, without section headers', () => {
    expectors.expectToMatchSnapshotWithData(data.MAP_DATA_LIST_ROWS);
  });

  it('renders Map: List rows, with section headers', () => {
    expectors.expectToMatchSnapshotWithData(data.MAP_DATA_LIST_ROWS, true);
  });

  it('renders Map: Map rows, without section headers', () => {
    expectors.expectToMatchSnapshotWithData(data.MAP_DATA_MAP_ROWS);
  });

  it('renders Map: Map rows, with section headers', () => {
    expectors.expectToMatchSnapshotWithData(data.MAP_DATA_MAP_ROWS, true);
  });

  it('renders basic Set', () => {
    expectors.expectToMatchSnapshotWithData(data.SET_DATA);
  });

  it('renders basic Range', () => {
    expectors.expectToMatchSnapshotWithData(data.RANGE_DATA);
  });
});

describe('ImmutableListView with delayed rendering', () => {
  it('renders basic List during interactions', () => {
    // Mock this method to make sure it's not run.
    InteractionManager.runAfterInteractions = () => {};

    const tree = renderer
      .create(
        <ImmutableListView
          immutableData={data.LIST_DATA}
          rowsDuringInteraction={1}
          renderRow={renderers.renderRow}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders basic List after interactions', () => {
    // Mock this method to make sure it runs immediately.
    InteractionManager.runAfterInteractions = (callback) => callback();

    const tree = renderer
      .create(
        <ImmutableListView
          immutableData={data.LIST_DATA}
          rowsDuringInteraction={1}
          renderRow={renderers.renderRow}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ImmutableListView with renderEmpty', () => {
  it('renders normally when there are some items', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.LIST_DATA}
        renderRow={renderers.renderRow}
        renderEmpty={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a function', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.EMPTY_DATA}
        renderRow={renderers.renderRow}
        renderEmpty={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a string', () => {
    const color = 'red';

    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.EMPTY_DATA}
        renderRow={renderers.renderRow}
        renderEmpty="No items"
        contentContainerStyle={{ color }}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("doesn't render empty with null", () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.EMPTY_DATA}
        renderRow={renderers.renderRow}
        renderEmpty={null}
        renderEmptyInList={null}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('ImmutableListView with renderEmptyInList', () => {
  it('renders normally when there are some items', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.LIST_DATA}
        renderRow={renderers.renderRow}
        renderEmptyInList={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a function', () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.EMPTY_DATA}
        renderRow={renderers.renderRow}
        renderEmptyInList={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a string', () => {
    const color = 'red';

    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.EMPTY_DATA}
        renderRow={renderers.renderRow}
        renderEmptyInList="No items"
        contentContainerStyle={{ color }}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("doesn't render empty with null", () => {
    const tree = renderer.create(
      <ImmutableListView
        immutableData={data.EMPTY_DATA}
        renderRow={renderers.renderRow}
        renderEmpty={null}
        renderEmptyInList={null}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
