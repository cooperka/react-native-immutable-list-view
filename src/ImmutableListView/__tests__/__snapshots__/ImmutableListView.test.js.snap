// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImmutableListView renders Map: List rows, with section headers 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": NaN,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={
    Array [
      0,
      4,
      6,
      7,
    ]
  }
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}
    >
      first (3 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "m"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "a"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "p"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}
    >
      second (1 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "foo"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}
    >
      third (0 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}
    >
      fourth (1 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "bar"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders Map: List rows, without section headers 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 4,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      ["m","a","p"]
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      ["foo"]
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      []
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      ["bar"]
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders Map: Map rows, with section headers 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": NaN,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={
    Array [
      0,
      3,
    ]
  }
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}
    >
      first (2 items)
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "data 1"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "data 2"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      header={true}
    >
      second (0 items)
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders Map: Map rows, without section headers 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 2,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      {"row1":"data 1","row2":"data 2"}
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      {}
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders basic List 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 3,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "lists"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "are"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "great"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders basic Range 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 3,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      3
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      6
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      9
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders basic Set 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 3,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "one"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "two"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "three"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders nested List 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 2,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      ["so","are"]
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      ["nested","lists"]
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView renders with empty data 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "padding": 8,
          "textAlign": "center",
        }
      }
    >
      No data.
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with delayed rendering renders basic List after interactions 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 3,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "lists"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "are"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "great"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with delayed rendering renders basic List during interactions 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "lists"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with renderEmpty doesn't render empty with null 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 0,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View />
</RCTScrollView>
`;

exports[`ImmutableListView with renderEmpty renders empty with a function 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  "No items"
</Text>
`;

exports[`ImmutableListView with renderEmpty renders empty with a string 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "padding": 8,
        "textAlign": "center",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  No items
</Text>
`;

exports[`ImmutableListView with renderEmpty renders normally when there are some items 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 3,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "lists"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "are"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "great"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with renderEmptyInList doesn't render empty with null 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 0,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View />
</RCTScrollView>
`;

exports[`ImmutableListView with renderEmptyInList renders empty with a function 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "No items"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with renderEmptyInList renders empty with a string 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  contentContainerStyle={
    Object {
      "color": "red",
    }
  }
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "padding": 8,
          "textAlign": "center",
        }
      }
    >
      No items
    </Text>
  </View>
</RCTScrollView>
`;

exports[`ImmutableListView with renderEmptyInList renders normally when there are some items 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 3,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "lists"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "are"
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "great"
    </Text>
  </View>
</RCTScrollView>
`;
